source/grid/grid_tools.cc

```diff
@@ -2726,16 +2752,42 @@ namespace GridTools
     // Get cell weighting if a signal has been attached to the triangulation
     if (!triangulation.signals.cell_weight.empty())
       {
+        // If an underlying shared::Tria allows artificial cells,
+        // then save the current set of subdomain ids, and set
+        // subdomain ids to the "true" owner of each cell. we later
+        // restore these flags.
+        const parallel::shared::Triangulation<dim, spacedim> *shared_tria =
+          (dynamic_cast<const parallel::shared::Triangulation<dim, spacedim> *>(
+            &triangulation));
+        std::vector<types::subdomain_id> saved_subdomain_ids;
+        if (shared_tria != nullptr && shared_tria->with_artificial_cells())
```

Comment:
same here. eventually, we should really introduce some helper function that just does the functionality of the whole block after this one :-)